Userland/Libraries/LibPDF/Renderer.cpp

```diff
@@ -701,7 +701,146 @@ RENDERER_HANDLER(inline_image_begin_data)
+{
+    // PDF 1.7 spec, 4.8.6 Inline Images:
+    // "Beginning with PDF 1.2, the value of the ColorSpace entry may also be the name
+    //  of a color space in the ColorSpace subdictionary of the current resource dictionary."
+
+    // But PDF 1.7 spec, 4.5.2 Color Space Families:
+    // "Outside a content stream, certain objects, such as image XObjects,
+    //  specify a color space as an explicit parameter, often associated with
+    //  the key ColorSpace. In this case, the color space array or name is
+    //  always defined directly as a PDF object, not by an entry in the
+    //  ColorSpace resource subdictionary."
+
+    // This converts a named color space of an inline image to an explicit color space object,
+    // so that the regular image drawing code tolerates it.
+
+    if (!color_space_value.has<NonnullRefPtr<Object>>())
+        return color_space_value;
+
+    auto const& object = color_space_value.get<NonnullRefPtr<Object>>();
+    if (!object->is<NameObject>())
+        return color_space_value;
+
+    auto const& name = object->cast<NameObject>()->name();
+    if (name == "DeviceGray" || name == "DeviceRGB" || name == "DeviceCMYK")
+        return color_space_value;
+
+    auto color_space_resource_dict = TRY(resources->get_dict(document, CommonNames::ColorSpace));
+    return color_space_resource_dict->get_object(document, name);
+}
+
+static PDFErrorOr<NonnullRefPtr<StreamObject>> expand_inline_image_abbreviations(NonnullRefPtr<StreamObject> inline_stream, NonnullRefPtr<DictObject> resources, RefPtr<Document> document)
+{
+    // TABLE 4.43 Entries in an inline image object
+    HashMap<DeprecatedFlyString, DeprecatedFlyString> key_expansions {
+        { "BPC", "BitsPerComponent" },
+        { "CS", "ColorSpace" },
+        { "D", "Decode" },
+        { "DP", "DecodeParms" },
+        { "F", "Filter" },
+        { "H", "Height" },
+        { "IM", "ImageMask" },
+        { "I", "Interpolate" },
+        { "Intent", "Intent" }, // "No abbreviation"
+        { "L", "Length" },      // PDF 2.0; would make more sense to read in Parser.
+        { "W", "Width" },
+    };
+
+    // TABLE 4.44 Additional abbreviations in an inline image object
+    // "Also note that JBIG2Decode and JPXDecode are not listed in Table 4.44
+    //  because those filters can be applied only to image XObjects."
+    HashMap<DeprecatedFlyString, DeprecatedFlyString> value_expansions {
+        { "G", "DeviceGray" },
+        { "RGB", "DeviceRGB" },
+        { "CMYK", "DeviceCMYK" },
+        { "I", "Indexed" },
+        { "AHx", "ASCIIHexDecode" },
+        { "A85", "ASCII85Decode" },
+        { "LZW", "LZWDecode" },
+        { "Fl", "FlateDecode" },
+        { "RL", "RunLengthDecode" },
+        { "CCF", "CCITTFaxDecode" },
+        { "DCT", "DCTDecode" },
+    };
+
+    // The values in key_expansions, that is the final expansions, are the valid keys in an inline image dict.
+    HashTable<DeprecatedFlyString> valid_keys;
```

Comment:
Any sense making these tables Function-statics?

seems a shame to init them for every image.

Reply:
These are so tiny and this happens so rarely that it doubt it matters. But sure, done.